### ISSUE TITLE
sDNA for QGIS 3.28 and 3.32 issue not running

### ISSUE THREAD

**Issue by alainjfc** [fiftysevendegreesofrad/sdna_open#15]
I recently installed a stable release of QGIS 3.28.11.1 (2023). sDNA was not running. Please take a look at the attached messages and the file. Previously I used QGIS 3.18.3 (2021). I also tried the latest version 3.32, and sDNA seems to have similar problems.

Please have a look when you have time. Let me know. Thank you. 
[M1_7_s.zip](https://github.com/fiftysevendegreesofrad/sdna_open/files/13072263/M1_7_s.zip)
[2023 10 QGIS 3.28.11 sDNA error.docx](https://github.com/fiftysevendegreesofrad/sdna_open/files/13072265/2023.10.QGIS.3.28.11.sDNA.error.docx)

**Comment by JamesParrott** [comment]
Hi Alain,

Issue #9 is causing an error here (time.clock being deprecated in Python 3).    There's a type error in PyShp too, but that could be because of Issue #9

These fixes aren't fully tested, but last year a couple of us successfully manually patched our sDNA installations using the following minor hacks.  If you don't need to run the Python tests, and will only be running sDNA using Python 3 (this will break it in Python 2 instead as there is no time.perf_counter in that version) the safest approach is:

 a) opening a file explorer admin window, open `c:\Program Files (x86)\sDNA\` and copy `runcalculation.py` to `runcalculation.old`.  Then open the first one (the .py) in a text editor as admin and replace `time.clock` with `time.perf_counter`.  Only two replacements are necessary, on line 167 and lin 169.  Save the file, and try sDNA again.

 b) a slightly more dangerous hack, but one that should work in both Python 2 and Python 3, and pass the test code, is to:
  opening a file explorer admin window, rename `c:\Program Files (x86)\sDNA\sdna_environment.py` to `c:\Program Files (x86)\sDNA\sdna_environment.old`, then download https://raw.githubusercontent.com/JamesParrott/sdna_open/Testing_time.clock_%3D_time.perf_counter/arcscripts/sdna_environment.py and copy in the new version od `sdna_environment.py`  This monkey patches the time module at the point where it is first imported (and calls self.close() on the shapefile Writer object when it's finished).

**Comment by JamesParrott** [comment]
An alternative interim solution is to try one of these experimental releases of sdna_plus: 
https://github.com/JamesParrott/sdna_plus/releases

**Comment by alainjfc** [comment]
The same problem occurs as well with ArcGIS Pro. Do you think this problem can be fixed at source?

**Comment by JamesParrott** [comment]
The open sourced sdna_plus main branch includes a similar Python 3 patch to `runcalcation.py` to the one I described above (using time.process_time, necessitating Python <=3.3) https://github.com/fiftysevendegreesofrad/sdna_plus/blob/9113e2504c09f3dd185832dbfcba793980c8abe6/arcscripts/runcalculation.py#L151.  

So in that sense, it has been fixed at source.  

If none of the 7 third-party-releases of sdna_plus above are suitable, or if you need something that I stripped out from them, let me know what is needed.  Otherwise, you must try the patching process above yourself after installation, or wait for an official release.

**Comment by alainjfc** [comment]
Hi James,

Thank you for the update
What are we waiting to make an official release?

Best
Alain

**Comment by JamesParrott** [comment]
You're welcome :).  I'm guessing all the stuff I stripped out in my fork (that anyone uses) has to be compiled too, to avoid a breaking change.  That's straightforward from what I've already got compiling in Visual Studio.  Confidence that it's been compiled -correctly- would be nice too, e.g. getting the tests running, and being passed, to be sure I haven't over looked an important compiler setting.  But it's not my call.
